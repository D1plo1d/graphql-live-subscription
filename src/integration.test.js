import {
  parse,
  subscribe,
  GraphQLSchema,
  GraphQLObjectType,
} from 'graphql'
import tql from 'typiql'

import {
  GraphQLLiveData,
  subscribeToLiveData,
} from './'

const getSubscriptionProvider = async (source, args, context, resolveInfo) => {
  return context.store
}

const getSource = async (originalSource, args, context, resolveInfo) => {
  return context.store.getState()
}

const HouseGraphQLType = new GraphQLObjectType({
  name: 'House',
  fields: () => ({
    id: {
      type: tql`ID!`
    },
    address: {
      type: tql`String!`
    },
    numberOfCats: {
      type: tql`Int!`
    },
    numberOfDogs: {
      type: tql`Int!`
    },
  }),
})

const type = () => tql`[${HouseGraphQLType}!]`

const schema = new GraphQLSchema({
  query: HouseGraphQLType,
  subscription: new GraphQLObjectType({
    name: 'SubscriptionRoot',
    fields: () => ({
      houses: {
        type: GraphQLLiveData({
          name: 'HouseLiveData',
          type,
        }),
        resolve(source) {
          return source
        },

        subscribe: subscribeToLiveData({
          getSubscriptionProvider,
          getSource,
          type,
        }),
      }
    })
  }),
})

const createTestSubscription = async () => {
  let onChangeInner
  const onChange = () => onChangeInner()
  const state = [
    {
      id: 'real_street',
      address: '123 real st',
      numberOfCats: 5,
      numberOfDogs: 7,
    },
    {
      id: 'legit_road',
      address: '200 legit rd',
      numberOfCats: 0,
      numberOfDogs: 1,
    },
  ]
  const store = {
    subscribe: (cb) => onChangeInner = cb,
    getState: () => state,
  }

  const document = parse(`
    subscription {
      houses {
        query {
          id
          address
          numberOfCats
        }
        patches {
          ... on RFC4627Add { op, path, value }
          ... on RFC4627Remove { op, path }
          ... on RFC4627Replace { op, path, value }
          ... on RFC4627Move { op, from, path }
          ... on RFC4627Copy { op, from, path }
          ... on RFC4627Test { op, path, value }
        }
      }
    }
  `)

  let subscription = subscribe({
    schema,
    document,
    contextValue: { store },
  })


  if (subscription.then != null) subscription = await subscription

  return {
    subscription,
    onChange,
    state,
  }
}

const expectSubscriptionResponse = async (subscription) => {
  const result = await subscription.next()

  expect(result.done).toEqual(false)
  expect(result.value.data).toMatchSnapshot()
}

describe('GraphQLLiveData Integration', () => {
  it('publishes the initialQuery immediately', async () => {
    const { subscription, onChange, state } = await createTestSubscription()
    expect(subscription.errors).toEqual(undefined)

    await expectSubscriptionResponse(subscription)
  })

  it('publishes diffs when the subscriber calls the fn passed to subscribe()', async () => {
    const { subscription, onChange, state } = await createTestSubscription()
    expect(subscription.errors).toEqual(undefined)
    // inital query
    await subscription.next()
    // first diff
    state[0].numberOfDogs = 0
    state[0].numberOfCats = 200
    onChange()
    await expectSubscriptionResponse(subscription)
    // second diff
    state[0].address = state[0].address + ' apt. 1'
    state[1].address = state[1].address + ' apt. 2'
    onChange()
    await expectSubscriptionResponse(subscription)
  })
})
