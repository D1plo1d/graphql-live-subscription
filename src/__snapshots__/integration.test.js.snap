// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLLiveData Integration publishes diffs when the subscriber calls the fn passed to subscribe() 1`] = `
Object {
  "houses": Object {
    "patches": Array [
      Object {
        "op": "replace",
        "path": "/0/numberOfCats",
        "value": 200,
      },
    ],
    "query": null,
  },
}
`;

exports[`GraphQLLiveData Integration publishes diffs when the subscriber calls the fn passed to subscribe() 2`] = `
Object {
  "houses": Object {
    "patches": Array [
      Object {
        "op": "replace",
        "path": "/1/address",
        "value": "200 legit rd apt. 2",
      },
      Object {
        "op": "replace",
        "path": "/0/address",
        "value": "123 real st apt. 1",
      },
    ],
    "query": null,
  },
}
`;

exports[`GraphQLLiveData Integration publishes the initialQuery immediately 1`] = `
Object {
  "houses": Object {
    "patches": null,
    "query": Array [
      Object {
        "address": "123 real st",
        "id": "real_street",
        "numberOfCats": 5,
      },
      Object {
        "address": "200 legit rd",
        "id": "legit_road",
        "numberOfCats": 0,
      },
    ],
  },
}
`;
